File: backend/app/services/clustering.py
from sklearn.cluster import DBSCAN
import numpy as np
import pandas as pd

def identify_blackspots(df, eps_km=0.1, min_samples=3):
    """
    Identifies accident blackspots using DBSCAN clustering.
    eps_km: maximum distance in km between two samples to be considered in the same cluster.
    """
    if df.empty or 'Latitude' not in df.columns or 'Longitude' not in df.columns:
        return []

    # Radius of Earth in km
    kms_per_radian = 6371.0088
    epsilon = eps_km / kms_per_radian

    # Filter out invalid coordinates
    valid_df = df.dropna(subset=['Latitude', 'Longitude'])
    if valid_df.empty:
        return []

    coords = valid_df[['Latitude', 'Longitude']].values
    
    # DBSCAN 
    db = DBSCAN(eps=epsilon, min_samples=min_samples, algorithm='ball_tree', metric='haversine').fit(np.radians(coords))
    
    valid_df = valid_df.copy()
    valid_df['Cluster'] = db.labels_
    
    # Filter out noise (label -1)
    clusters = valid_df[valid_df['Cluster'] != -1]
    
    if clusters.empty:
        return []

    # Summarize clusters
    blackspots = []
    for cluster_id, group in clusters.groupby('Cluster'):
        # Calculate severity score: Fatal*5 + Grievous*3 + Minor*1
        # Match severity strings from dataset if possible, or use heuristic
        fatal = len(group[group['Severity'].str.contains('Fatal', case=False, na=False)])
        grievous = len(group[group['Severity'].str.contains('Grievous', case=False, na=False)])
        minor = len(group) - fatal - grievous
        
        score = (fatal * 5) + (grievous * 3) + (minor * 1)
        
        blackspots.append({
            "id": int(cluster_id),
            "location": f"Cluster {cluster_id} ({group['District'].iloc[0]})",
            "accidents": len(group),
            "fatalities": fatal,
            "score": round(score / len(group), 2),
            "lat": group['Latitude'].mean(),
            "lng": group['Longitude'].mean()
        })
    
    # Sort by score descending
    blackspots.sort(key=lambda x: x['score'], reverse=True)
    return blackspots

def get_heatmap_data(df):
    """
    Returns points for heatmap and frontend clustering.
    """
    if df.empty or 'Latitude' not in df.columns or 'Longitude' not in df.columns:
        return []
        
    cols = ['Latitude', 'Longitude', 'Severity', 'District']
    available_cols = [c for c in cols if c in df.columns]
    
    points = df[available_cols].copy()
    points.rename(columns={'Latitude': 'lat', 'Longitude': 'lng', 'Severity': 'severity', 'District': 'district'}, inplace=True)
    points['weight'] = 1.0
    return points.to_dict(orient='records')

File: frontend/src/services/clustering.js
/**
 * DBSCAN implementation in JavaScript
 * For spatial clustering of accident points.
 */

// Haversine distance in meters
function getDistance(p1, p2) {
  const R = 6371000; // Earth radius in meters
  const dLat = (p2.lat - p1.lat) * Math.PI / 180;
  const dLng = (p2.lng - p1.lng) * Math.PI / 180;
  const a = Math.sin(dLat / 2) * Math.sin(dLat / 2) +
            Math.cos(p1.lat * Math.PI / 180) * Math.cos(p2.lat * Math.PI / 180) *
            Math.sin(dLng / 2) * Math.sin(dLng / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return R * c;
}

export function runDBSCAN(points, eps, minSamples) {
  const n = points.length;
  const labels = new Array(n).fill(null); // null = unvisited, -1 = noise, 0+ = cluster ID
  let clusterId = 0;

  for (let i = 0; i < n; i++) {
    if (labels[i] !== null) continue;

    const neighbors = getNeighbors(points, i, eps);

    if (neighbors.length < minSamples) {
      labels[i] = -1; // Noise
    } else {
      expandCluster(points, i, neighbors, clusterId, eps, minSamples, labels);
      clusterId++;
    }
  }

  // Summarize clusters
  const clusters = [];
  for (let i = 0; i < clusterId; i++) {
    const clusterPoints = points.filter((_, idx) => labels[idx] === i);
    
    // Calculate metrics
    const latSum = clusterPoints.reduce((sum, p) => sum + p.lat, 0);
    const lngSum = clusterPoints.reduce((sum, p) => sum + p.lng, 0);
    const count = clusterPoints.length;
    
    const fatalities = clusterPoints.filter(p => 
      p.severity && p.severity.toLowerCase().includes('fatal')
    ).length;
    
    const grievous = clusterPoints.filter(p => 
      p.severity && p.severity.toLowerCase().includes('grievous')
    ).length;
    
    const minor = count - fatalities - grievous;
    const score = (fatalities * 5 + grievous * 3 + minor * 1) / count;

    clusters.push({
      id: i,
      location: `Cluster ${i} (${clusterPoints[0].district || 'Unknown'})`,
      accidents: count,
      fatalities: fatalities,
      score: parseFloat(score.toFixed(2)),
      lat: latSum / count,
      lng: lngSum / count,
      points: clusterPoints
    });
  }

  return {
    clusters: clusters.sort((a, b) => b.score - a.score),
    labels: labels
  };
}

function getNeighbors(points, pointIdx, eps) {
  const neighbors = [];
  for (let i = 0; i < points.length; i++) {
    if (getDistance(points[pointIdx], points[i]) <= eps) {
      neighbors.push(i);
    }
  }
  return neighbors;
}

function expandCluster(points, pointIdx, neighbors, clusterId, eps, minSamples, labels) {
  labels[pointIdx] = clusterId;
  let i = 0;
  while (i < neighbors.length) {
    const neighborIdx = neighbors[i];
    
    if (labels[neighborIdx] === -1) {
      labels[neighborIdx] = clusterId;
    } else if (labels[neighborIdx] === null) {
      labels[neighborIdx] = clusterId;
      const nextNeighbors = getNeighbors(points, neighborIdx, eps);
      if (nextNeighbors.length >= minSamples) {
        neighbors.push(...nextNeighbors.filter(idx => !neighbors.includes(idx)));
      }
    }
    i++;
  }
}
